lib/CleanPlugin.js

```diff
@@ -0,0 +1,224 @@
+const schema = require("../schemas/plugins/CleanPlugin.json");
+const { join } = require("./util/fs");
+
+/** @typedef {import("../declarations/plugins/CleanPlugin").CleanPluginArgument} CleanPluginArgument */
+/** @typedef {import("./Compiler")} Compiler */
+
+/** @typedef {(function(string):boolean)|RegExp} IgnoreItem */
+/** @typedef {function(IgnoreItem): void} AddToIgnoreCallback */
+
+/**
+ * @typedef {Object} CleanPluginCompilationHooks
+ * @property {SyncHook<[AddToIgnoreCallback]>} ignore
+ */
+
+/** @type {WeakMap<Compilation, CleanPluginCompilationHooks>} */
+const compilationHooksMap = new WeakMap();
+
+/**
+ * @param {IgnoreItem} ignore regexp or function
+ * @param {string} asset asset path
+ * @returns {boolean} true if an asset should be ignored
+ */
+const checkToIgnore = (ignore, asset) => {
+	if (ignore instanceof RegExp) {
+		return !!ignore.exec(asset);
+	}
+
+	return ignore(asset);
+};
+
+class CleanPlugin {
+	/**
+	 * @param {Compilation} compilation the compilation
+	 * @returns {CleanPluginCompilationHooks} the attached hooks
+	 */
+	static getCompilationHooks(compilation) {
+		if (!(compilation instanceof Compilation)) {
+			throw new TypeError(
+				"The 'compilation' argument must be an instance of Compilation"
+			);
+		}
+		let hooks = compilationHooksMap.get(compilation);
+		if (hooks === undefined) {
+			hooks = {
+				/** @type {SyncHook<[AddToIgnoreCallback]>} */
+				ignore: new SyncHook(["ignore"])
+			};
+			compilationHooksMap.set(compilation, hooks);
+		}
+		return hooks;
+	}
+
+	/** @param {CleanPluginArgument} [options] options */
+	constructor(options) {
+		this.options = Object.assign({ enabled: true, dry: false }, options || {});
+
+		if (options && typeof options === "object") {
+			validate(schema, options, {
+				name: "Clean Plugin",
+				baseDataPath: "options"
+			});
+		}
+
+		/** @type {IgnoreItem[]} */
+		this.ignoreList = [];
+
+		/** @type {import("./logging/Logger").Logger} */
+		this.logger = null;
+
+		/** @type {import("./util/fs").OutputFileSystem} */
+		this.fs = null;
+
+		/** @type {{files: Set<string>, directories: Set<string>}} */
+		this.fsState = {
+			files: new Set(),
+			directories: new Set()
+		};
+
+		if (this.options.ignore) {
+			this.ignoreList.push(this.options.ignore);
+		}
+	}
+
+	/**
+	 * @param {Compiler} compiler the compiler instance
+	 * @returns {void}
+	 */
+	apply(compiler) {
+		/**
+		 * @param {IgnoreItem} item regexp or function
+		 * @returns {void}
+		 */
+		const ignoreFn = item => void this.ignoreList.push(item);
+		const handleAsset = asset => {
+			const parts = asset.split(/[\\/]+/).slice(0, -1);
+
+			this.fsState.files.add(asset);
+			parts.reduce((all, part) => {
+				const directory = path.join(all, part);
+				this.fsState.directories.add(directory);
+				return directory;
+			}, "");
+		};
+		this.logger = compiler.getInfrastructureLogger("webpack.Clean");
+		this.fs = compiler.outputFileSystem;
+
+		compiler.hooks.emit.tapAsync(
```

Comment:
Not sure it is safe to use `emit`, because I see a lot of plugins with `stage: 99999`/any number and using emit, maybe better introduce new hook, but let's wait @sokra answer

Reply:
I think that a new hook will not solve this problem, because any 3-rd party plugin may use this new hook too )

Reply:
We can implement hook `cleanEmittedAssets`, so nobody will use it for emitting :smile:

Reply:
But they can )))